File: internal/infra/astscan/pyscan/scanner.go
// Package pyscan implements astscan.LanguageScanner for Python using
// tree-sitter with the tree-sitter-python grammar. Detection patterns
// mirror the JS scanner's design — see jsscan/queries.scm for the
// rationale; each "@cap.X" capture maps to a domain.Capability via
// capabilityFor() below.
//
// Concurrency: a Scanner holds an immutable Query that can be shared
// across goroutines for read. Parsers and QueryCursors are created
// per-call (cheap) but cursors are pooled when scanning lots of files
// in parallel (matches the snapshot enrich worker pool of 8).
//
// Coverage today (mirrors jsscan capability set):
//
//   - shell-spawn         subprocess.{run,Popen,...}, os.{system,popen,exec*}, pty.spawn
//   - dynamic-eval        eval, exec, compile, __import__
//   - base64-decode       base64/codecs/binascii decode functions
//   - net-egress          urllib/requests/httpx/aiohttp/socket/http.client
//   - env-read            os.environ['X'] / os.environ.get('X') / os.getenv('X')
//   - fs-write-outside-root open(...'w'/'a'), pathlib write_*, shutil copy/move
//   - raw-ip-literal      http://NNN.NNN.NNN.NNN string literals
//
// install-hook is detected manifest-side (heuristics package handles
// setup.py / pyproject.toml install hooks separately when those parsers
// land). Source-pattern heuristics (obfuscated payload, suspicious URL)
// are also separate — they're regex-over-source and language-agnostic.
package pyscan

import (
	_ "embed"
	"strings"
	"sync"

	"github.com/qwexvf/aegis-cli/internal/domain"
	"github.com/qwexvf/aegis-cli/internal/infra/astscan"

	ts "github.com/tree-sitter/go-tree-sitter"
	tspython "github.com/tree-sitter/tree-sitter-python/bindings/go"
)

//go:embed queries.scm
var queriesSource string

// Scanner is the Python analyzer. Construct via New; the heavy lifting
// (compiling the language and queries) happens once at startup.
type Scanner struct {
	lang  *ts.Language
	query *ts.Query

	// captureToCap maps capture index → Capability. Pre-computed so
	// the per-match hot loop avoids string lookups.
	captureToCap map[uint]domain.Capability

	// envReadCaptureIdx is the capture index of @env_var (the actual
	// env-name token). -1 if not present in the query.
	envReadCaptureIdx int

	cursors sync.Pool
}

// New compiles the language and queries. Returns an error if the
// embedded query string is malformed (a developer bug — should never
// happen in shipping binaries).
func New() (*Scanner, error) {
	lang := ts.NewLanguage(tspython.Language())
	q, qerr := ts.NewQuery(lang, queriesSource)
	if qerr != nil {
		return nil, qerr
	}

	captureToCap := map[uint]domain.Capability{}
	envIdx := -1
	for i, name := range q.CaptureNames() {
		idx := uint(i)
		switch {
		case strings.HasPrefix(name, "cap."):
			if c := capabilityFor(name); c != 0 {
				captureToCap[idx] = c
			}
		case name == "env_var":
			envIdx = i
		}
	}

	return &Scanner{
		lang:              lang,
		query:             q,
		captureToCap:      captureToCap,
		envReadCaptureIdx: envIdx,
		cursors: sync.Pool{New: func() any {
			return ts.NewQueryCursor()
		}},
	}, nil
}

// AnalyzeFile satisfies astscan.LanguageScanner. Walks the AST of the
// given Python source bytes, fires every capability whose query
// matches, and accumulates env-var reads for the credential-name
// filter applied at scoring time.
func (s *Scanner) AnalyzeFile(path string, body []byte, f *astscan.Findings) {
	parser := ts.NewParser()
	defer parser.Close()
	// SetLanguage can fail only if the language ABI version mismatches
	// the parser binary; both are vendored at compile time, so this is
	// effectively impossible — but explicit discard documents intent.
	_ = parser.SetLanguage(s.lang)

	tree := parser.Parse(body, nil)
	if tree == nil {
		return
	}
	defer tree.Close()

	cursor := s.cursors.Get().(*ts.QueryCursor)
	defer func() {
		s.cursors.Put(cursor)
	}()

	matches := cursor.Matches(s.query, tree.RootNode(), body)
	for {
		m := matches.Next()
		if m == nil {
			break
		}
		for _, cap := range m.Captures {
			if c, ok := s.captureToCap[uint(cap.Index)]; ok {
				f.AddCapability(c)
				if f.CollectEvidence {
					line := int(cap.Node.StartPosition().Row) + 1
					f.AddEvidence(c, path, line, string(cap.Node.Utf8Text(body)))
				}
				continue
			}
			if int(cap.Index) == s.envReadCaptureIdx {
				name := string(cap.Node.Utf8Text(body))
				f.AddEnvRead(name)
			}
		}
	}
}

// capabilityFor maps a "cap.XXX" capture name to the Capability enum.
// Names match Capability.String() suffixes for readability.
func capabilityFor(captureName string) domain.Capability {
	suffix := strings.TrimPrefix(captureName, "cap.")
	for _, c := range domain.AllCapabilities() {
		if c.String() == suffix {
			return c
		}
	}
	return 0
}

File: crates/aegis-ast/src/scanner.rs
//! Generic tree-sitter capability scanner shared by every language.
//! Port of the common logic in each `ast/<lang>/scanner.go` (the part
//! that's identical across languages — `@cap.<name>` → Capability and
//! `@env_var` → env-read). Language-specific extras (JS taint pass) live
//! in the per-language module.

use aegis_domain::{Capability, ALL_CAPABILITIES};
use streaming_iterator::StreamingIterator;
use tree_sitter::{Language, Parser, Query, QueryCursor};

use crate::{Findings, LanguageScanner};

/// An optional second pass run on the same parse tree after the query
/// pass (e.g. the JS constant-fold/taint analysis, which tree-sitter
/// queries alone can't express).
pub type PostPass = for<'a> fn(tree_sitter::Node<'a>, &[u8], &str, &mut Findings);

/// A compiled grammar + query. Cheap to share; the heavy compile is
/// one-time. Mirrors the common `Scanner` shape.
pub struct GrammarScanner {
    language: Language,
    query: Query,
    /// capture index → capability (pre-computed; hot loop avoids lookups).
    capture_to_cap: Vec<Option<Capability>>,
    /// capture index of `@env_var`, if present.
    env_read_idx: Option<u32>,
    /// optional language-specific second pass over the same tree.
    post_pass: Option<PostPass>,
}

impl GrammarScanner {
    /// Compile `query_src` against `language`. Errors on a malformed
    /// query (a developer bug — the query is embedded at build time).
    pub fn new(language: Language, query_src: &str) -> Result<Self, String> {
        Self::with_post_pass(language, query_src, None)
    }

    /// Like [`new`](Self::new) but attaches a second pass over the tree.
    pub fn with_post_pass(
        language: Language,
        query_src: &str,
        post_pass: Option<PostPass>,
    ) -> Result<Self, String> {
        let query = Query::new(&language, query_src).map_err(|e| format!("query: {e}"))?;
        let names = query.capture_names();
        let mut capture_to_cap = vec![None; names.len()];
        let mut env_read_idx = None;
        for (i, name) in names.iter().enumerate() {
            if let Some(suffix) = name.strip_prefix("cap.") {
                capture_to_cap[i] = capability_for(suffix);
            } else if *name == "env_var" {
                env_read_idx = Some(i as u32);
            }
        }
        Ok(GrammarScanner {
            language,
            query,
            capture_to_cap,
            env_read_idx,
            post_pass,
        })
    }
}

impl LanguageScanner for GrammarScanner {
    fn analyze_file(&self, path: &str, body: &[u8], findings: &mut Findings) {
        let mut parser = Parser::new();
        if parser.set_language(&self.language).is_err() {
            return;
        }
        let Some(tree) = parser.parse(body, None) else {
            return;
        };

        let mut cursor = QueryCursor::new();
        let mut matches = cursor.matches(&self.query, tree.root_node(), body);
        while let Some(m) = matches.next() {
            for cap in m.captures {
                let idx = cap.index as usize;
                if let Some(Some(c)) = self.capture_to_cap.get(idx) {
                    findings.add_capability(*c);
                    if findings.collect_evidence {
                        let line = cap.node.start_position().row + 1;
                        let snippet = cap.node.utf8_text(body).unwrap_or("").to_string();
                        findings.add_evidence(*c, path, line, snippet);
                    }
                } else if Some(cap.index) == self.env_read_idx {
                    if let Ok(name) = cap.node.utf8_text(body) {
                        findings.add_env_read(name);
                    }
                }
            }
        }

        // Optional language-specific second pass over the same tree.
        if let Some(pp) = self.post_pass {
            pp(tree.root_node(), body, path, findings);
        }
    }
}

/// Map a `cap.<name>` suffix to a [`Capability`] by its canonical name.
/// Mirrors `capabilityFor`.
fn capability_for(suffix: &str) -> Option<Capability> {
    ALL_CAPABILITIES
        .iter()
        .copied()
        .find(|c| c.name() == suffix)
}
